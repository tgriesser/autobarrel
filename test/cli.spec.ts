import * as path from "path"
import * as fs from "fs-extra"
import childProcess from "child_process"
import glob from "glob"
import util from "util"
import pDefer from "p-defer"

const globAsync = util.promisify(glob)

beforeEach(() => {
  try {
    fs.rmdirSync(path.join(__dirname, "testing"), { recursive: true })
  } catch {}
  fs.copySync(path.join(__dirname, "fixtures"), path.join(__dirname, "testing"))
})

afterEach(() => {
  try {
    fs.rmdirSync(path.join(__dirname, "testing"), { recursive: true })
  } catch {}
})

describe("autobarrel cli", () => {
  test("should run the CLI", async () => {
    const result = childProcess.spawnSync(
      "./cli.js",
      ["--config", path.join(__dirname, "testing", "autobarrel.json")],
      {
        cwd: path.join(__dirname, "../dist"),
        stdio: "inherit",
      }
    )
    const globbed = await globAsync(path.join(__dirname, "testing/**"))
    expect(globbed).toMatchSnapshot()
  })
})

describe("watch mode", () => {
  const watchArgs = [
    "--config",
    path.join(__dirname, "testing", "autobarrel.json"),
    "--watch",
  ]

  // Run the watcher
  let spawned: childProcess.ChildProcess
  beforeEach(async () => {
    const dfd = pDefer()
    spawned = childProcess.spawn("./cli.js", watchArgs, {
      cwd: path.join(__dirname, "../dist"),
    })
    spawned.stdout?.on("data", (chunk) => {
      if (String(chunk).includes("Autobarrel watching")) {
        dfd.resolve()
      } else {
        console.log(String(chunk))
      }
    })
    spawned.stderr?.pipe(process.stderr)
    await dfd.promise
  })

  afterEach(() => {
    spawned.kill("SIGKILL")
    spawned.removeAllListeners()
  })

  const readWatchIndex = async () => {
    return await fs.readFile(
      path.join(__dirname, "testing/testDir/watchTest/index.ts"),
      "utf8"
    )
  }

  const delay = (ms: number) => {
    return new Promise((resolve) => setTimeout(resolve, ms))
  }

  test("should run --watch mode, watch file add", async () => {
    expect(await readWatchIndex()).toMatchSnapshot("Before file add")
    await fs.writeFile(
      path.join(__dirname, "testing/testDir/watchTest/added.ts"),
      "export const added = {}"
    )
    await delay(1000)
    expect(await readWatchIndex()).toMatchSnapshot("After file add")
  })

  test("should run --watch mode, watch file remove", async () => {
    expect(await readWatchIndex()).toMatchSnapshot("Before file remove")
    await fs.unlink(path.join(__dirname, "testing/testDir/watchTest/watch.ts"))
    await delay(1000)
    expect(readWatchIndex()).rejects.toHaveProperty(
      "message",
      expect.stringContaining("ENOENT: no such file or directory")
    )
  })

  test("should run --watch mode, watch dir add", async () => {
    expect(await readWatchIndex()).toMatchSnapshot("Before watch")
    await fs.mkdirp(path.join(__dirname, "testing/testDir/watchTest/addedDir"))
    await fs.writeFile(
      path.join(__dirname, "testing/testDir/watchTest/addedDir/watch.ts"),
      "export const addedDir = {}"
    )
    expect(await readWatchIndex()).toMatchSnapshot("After watch")
  })

  describe("dir remove", () => {
    beforeEach(async () => {
      await fs.mkdirp(
        path.join(__dirname, "testing/testDir/watchTest/addedDir")
      )
      await fs.writeFile(
        path.join(__dirname, "testing/testDir/watchTest/addedDir/watch.ts"),
        "export const addedDir = {}"
      )
      await delay(1000)
    })

    test("should run --watch mode, watch dir remove", async () => {
      expect(await readWatchIndex()).toMatchSnapshot("Before dir remove")
      await fs.remove(
        path.join(__dirname, "testing/testDir/watchTest/addedDir")
      )
      await delay(1000)
      expect(await readWatchIndex()).toMatchSnapshot("After dir remove")
    })
  })
})
