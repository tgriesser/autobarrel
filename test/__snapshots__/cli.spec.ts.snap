// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`autobarrel cli should run the CLI 1`] = `
Array [
  "testing",
  "testing/autobarrel.json",
  "testing/README.md",
  "testing/testDir",
  "testing/testDir/excludedDir",
  "testing/testDir/excludedDir/excluded.ts",
  "testing/testDir/excludedDir/index.ts",
  "testing/testDir/excludedFile",
  "testing/testDir/excludedFile/excludedFile.ts",
  "testing/testDir/excludedFile/includedFile.ts",
  "testing/testDir/excludedFile/index.ts",
  "testing/testDir/excludedPattern",
  "testing/testDir/excludedPattern/abc.bin.ts",
  "testing/testDir/excludedPattern/def.bin.ts",
  "testing/testDir/excludedPattern/index.ts",
  "testing/testDir/excludedPattern/some.gen.ts",
  "testing/testDir/excludeDts",
  "testing/testDir/excludeDts/index.ts",
  "testing/testDir/excludeDts/nonDts.ts",
  "testing/testDir/excludeDts/something.d.ts",
  "testing/testDir/ignoredDir",
  "testing/testDir/ignoredDir/ignoredDir.ts",
  "testing/testDir/index.ts",
  "testing/testDir/jsTs",
  "testing/testDir/jsTs/file.js",
  "testing/testDir/jsTs/index.ts",
  "testing/testDir/jsTs/tsFile.ts",
  "testing/testDir/noTsFiles",
  "testing/testDir/noTsFiles/README.md",
  "testing/testDir/some",
  "testing/testDir/some/deep",
  "testing/testDir/some/deep/index.ts",
  "testing/testDir/some/deep/path",
  "testing/testDir/some/deep/path/file.ts",
  "testing/testDir/some/deep/path/index.ts",
  "testing/testDir/some/index.ts",
  "testing/testDir/someDir",
  "testing/testDir/someDir/app.ts",
  "testing/testDir/someDir/index.ts",
  "testing/testDir/watchTest",
  "testing/testDir/watchTest/index.ts",
  "testing/testDir/watchTest/watch.ts",
]
`;

exports[`watch mode dir remove should run --watch mode, watch dir remove: After dir remove 1`] = `
"// created by autobarrel, do not modify directly

export * from './watch'
"
`;

exports[`watch mode dir remove should run --watch mode, watch dir remove: Before dir remove 1`] = `
"// created by autobarrel, do not modify directly

export * from './addedDir/'
export * from './watch'
"
`;

exports[`watch mode should run --watch mode, watch file add: After file add 1`] = `
"// created by autobarrel, do not modify directly

export * from './added'
export * from './watch'
"
`;

exports[`watch mode should run --watch mode, watch file add: Before file add 1`] = `
"// created by autobarrel, do not modify directly

export * from './watch'
"
`;

exports[`watch mode should run --watch mode, watch file remove: Before file remove 1`] = `
"// created by autobarrel, do not modify directly

export * from './watch'
"
`;
